### DISCUSSION TITLE
The width evalutating result of right quotation mark (’) of "Noto Sans CJK TC" is smaller than its really value.

### DISCUSSION THREAD

**Discussion by Yoxem** [harfbuzz/harfbuzz#3203] [Q&A] [answered]
I want to evaluating the width of a phrase "d'aoís" in Irish to layout it which contains right quotation mark U+2019 with [Google Noto Sans CJK TC font](https://www.google.com/get/noto/help/cjk/).

Howerver, its rendering result in Pango is full width like the screenshot:
![圖片](https://user-images.githubusercontent.com/184107/132981446-936344dc-c4c9-426a-9740-4286d200dc65.png)

but the x_advance of it (gid726) is 3.34375(see below), and even the half-width "d" is wider than it.

How to solve the problem? or does the font have another font face that I have to specify to make it evaluate correctly?

    ["d’aoís"]
    gid69[None]=0@12.09375:0.0,     0.0+0.0
    gid726[None]=1@3.34375:0.0,     0.0+0.0
    gid66[None]=4@10.9375:0.0,      0.0+0.0
    gid80[None]=5@11.875:0.0,       0.0+0.0
    gid171[None]=6@5.125:0.0,       0.0+0.0
    gid84[None]=8@9.03125:0.0,      0.0+0.0

**Comment by khaledhosny** [comment]
That is a question better asked to Noto CJK developers, HarfBuzz is not doing anything special here.

**Comment by jfkthame** [comment, marked as answer]
It's language-specific behavior in the font. The result depends what language code (if any) is passed to the shaper:

```
$ ./util/hb-shape NotoSansCJK-Regular.ttc  --language en --unicodes 41,2019,41
[gid34=0+544|gid719=1+213|gid34=2+607]

$ ./util/hb-shape NotoSansCJK-Regular.ttc  --language ja --unicodes 41,2019,41
[gid34=0+607|gid63015=1+279|gid34=2+607]

$ ./util/hb-shape NotoSansCJK-Regular.ttc  --language zh --unicodes 41,2019,41
[gid34=0+607|gid63154=1+1000|gid34=2+607]
```

So when the text is tagged as Japanese, the font substitutes an apostrophe that is slightly wider than the English one (actually, the main width difference arises because it doesn't kern with 'A'; it's also at a different height, as it happens); and for Chinese, it substitutes a fullwidth glyph.

**Reply by khaledhosny** [reply]
The language-specific subsets do this even without any language set.

**Reply by Yoxem** [reply]
I think that pango will choose the user's default locale to demo the language subset. I use a `zh` locale, so Pango uses it (and I found that the other language tags (eg. `ja`, `en`) set to the context of Pango doesn't have any effect on it, but it's another story).
